File: Yocto/poky/meta-mylayer/recipes-example/piracer/files/gamepad.py
# Copyright (C) 2022 twyleg
from piracer.vehicles import PiRacerStandard, PiRacerPro
from piracer.gamepads import ShanWanGamepad

if __name__ == '__main__':

    shanwan_gamepad = ShanWanGamepad()
    # piracer = PiRacerPro()
    piracer = PiRacerStandard()

    while True:
        gamepad_input = shanwan_gamepad.read_data()

        throttle = gamepad_input.analog_stick_right.y * 0.5
        steering = -gamepad_input.analog_stick_left.x
        
        print(f'throttle={throttle}, steering={steering}')
       
        if gamepad_input.button_y:
            print("Y pressed")
        if gamepad_input.button_x:
            print("X pressed")
        if gamepad_input.button_a:
            print("A pressed")
        if gamepad_input.button_b:
            print("B pressed")

        piracer.set_throttle_percent(throttle)
        piracer.set_steering_percent(steering)

File: Yocto/poky/meta-mylayer/recipes-app/piracerservice/files/src/GamePad.cpp
#include "GamePad.hpp"
#include <iostream>

GamePad::GamePad()
{
    Py_Initialize();
    pModule = PyImport_ImportModule("piracer.gamepads");
    pClass = PyObject_GetAttrString(pModule, "ShanWanGamepad");
    pInstance = PyObject_CallObject(pClass, NULL);
    
    throttle = 0.0;
    steering = 0.0;
    
    button_p = false;
    button_r = false;
    button_n = false;
    button_d = false;
}

GamePad::~GamePad()
{
    Py_DECREF(pThrottle);
    Py_DECREF(pSteering);
    Py_DECREF(pInput);
    Py_DECREF(pInstance);
    Py_DECREF(pClass);
    Py_DECREF(pModule);
    Py_Finalize();
}

// Read control input from the gamepad
void GamePad::readControl()
{
    // Call the "read_data" method on the gamepad instance
    pInput = PyObject_CallMethod(pInstance, "read_data", NULL);
    
    pThrottle = PyObject_GetAttrString(pInput, "analog_stick_right");
    pThrottle = PyObject_GetAttrString(pThrottle, "y");
    throttle = PyFloat_AsDouble(pThrottle);
    
    pSteering = PyObject_GetAttrString(pInput, "analog_stick_left");
    pSteering = PyObject_GetAttrString(pSteering, "x");
    steering = PyFloat_AsDouble(pSteering);
    
    button_p = PyObject_IsTrue(PyObject_GetAttrString(pInput, "button_y"));
    button_r = PyObject_IsTrue(PyObject_GetAttrString(pInput, "button_x"));
    button_n = PyObject_IsTrue(PyObject_GetAttrString(pInput, "button_a"));
    button_d = PyObject_IsTrue(PyObject_GetAttrString(pInput, "button_b"));
}

// Get the current throttle value
double GamePad::getThrottle()
{
    return throttle;
}

// Get the current steering value
double GamePad::getSteering()
{
    return steering;
}

bool GamePad::getButtonP()
{
    return button_p;
}

bool GamePad::getButtonR()
{
    return button_r;
}

bool GamePad::getButtonN()
{
    return button_n;
}

bool GamePad::getButtonD()
{
    return button_d;
}
